Add an API to manage projects, and use it pull their Docker images.

jankeromnes opened the pull request.

jankeromnes commented:
Potential problem with this implementation: Pull API is public, and can be spammed by anyone. Maybe we should rate limite?

nt1m commented on api/projects-api.js:
```diff
@@ -0,0 +1,65 @@
+  description: 'Trigger a Docker image pull for a given software project.',
+
+  handler: (request, response) => {
+    const { project: projectId } = request.query;
```
nit: Probably more simple to do:
```
const projectId = request.query.project;
```

nt1m commented on api/projects-api.js:
```diff
@@ -0,0 +1,65 @@
+    const { project: projectId } = request.query;
+    const projects = db.get('projects');
+    if (!projects[projectId]) {
+      response.statusCode = 404;
```
nit: `response.statusCode = 404; // Not found`

nt1m commented on api/projects-api.js:
```diff
@@ -0,0 +1,65 @@
+    machines.pull(projectId, error => {
+      if (error) {
+        log('[fail] pulling project', projectId, error);
+        response.statusCode = 500;
```
nit: `response.statusCode = 500; // Internal Error`

nt1m commented on api/projects-api.js:
```diff
@@ -0,0 +1,65 @@
+  title: 'Pull a project',
+  description: 'Trigger a Docker image pull for a given software project.',
+
+  handler: (request, response) => {
```
This API shouldn't be public.

Maybe limit this to logged in users only, or even just admins (`users.isAdmin(user)`)

jankeromnes replied:
The thing is that I'd also like to support Docker Hub web hooks. Maybe there is a way to authenticate that as well, eg with a shared secret.

nt1m replied:
So, maybe using authentication would be a good choice. Or maybe just rate limit the API.

jankeromnes replied:
Hm, I'd love to make this API public somehow, or at least very easily accessible to Docker Hub web hooks, but let's do that in another pull request. I'll make this handler admin-only to begin with.

jankeromnes commented:
I added a `docker.inspectImage()` method in order to verify the creation date of the image.
- If the image is **more recent** than our current production image, we tag it with `:janitor-production` (and set `project.data.updated` to the creation date, not the pull date)
- If the image is **as old as, or older than** our current production image (e.g. we already pulled this image, or we have an incremental update image that is more recent), then we still keep it, but we don't tag it with `:janitor-production` (and we don't change `project.data.updated`).

nt1m commented on api/projects-api.js:
```diff
@@ -0,0 +1,75 @@
+      return;
+    }
+
+    const projectId = request.query.project;
```
nit: above, use `const { user, query } = request;` and here: `const projectId = query.project;`

nt1m commented on api/projects-api.js:
```diff
@@ -0,0 +1,75 @@
+        return;
+      }
+
+      response.json({}, null, 2);
```
Not sure what's the point of this response. Why not just response.end() ?

jankeromnes commented:
Oh, so clicking on "Update branch" creates a merge commit... 

![giphy](https://user-images.githubusercontent.com/599268/28710401-10146f86-7384-11e7-98b6-a16bec4af3e1.gif)
